apt_repository: modify comment line in sources.list

pviffx opened the pull request:
##### SUMMARY

<!--- Describe the change below, including rationale and design decisions -->

Rebuild the SourceList from its parsed chunks. Use re.split() instead of split(), so it will not make additional spaces disappear. Furthermore, join the comment and source lines without additional spaces before and after the hash.

<!--- Add "Fixes #1234" or steps to reproduce the problem if there is no corresponding issue -->
Fixes #54403

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION  
This PR is a copy of #82087 that got closed because of inactivity from the original contributer.

Commit:
Fix for 'apt_repository modify comment line /etc/apt/sources.list' (#54403)

Rebuild the SourceList from its parsed chunks. Use re.split() instead of split(), so it will not make additional spaces disappear. Furthermore, join the comment and source lines without additional spaces before and after the hash.

Fixes #54403

sndrr committed:
Create 54403-apt_repository.yml

Add changelog fragment

bcoca reviewed (changes requested):
you can also copy the previous changelog, but what we really want is tests to avoid regressions

pviffx commented:
I finally managed to run tests following [Ansible dev guide](https://docs.ansible.com/ansible/latest/dev_guide/testing.html#developing-testing), some fails on units tests but they still fail on `devel` and are unrelated to apt_repository module, aside from that  sanity and integration tests have passed.    

I also tested on a fresh Deb12 with Ansible 2.17.9 with the following steps:  
- Install docker and its apt repo with `wget -O- https://get.docker.com | sudo bash` and add some comments to newly created `/etc/apt/sources.list.d/docker.list`.  
- Run an Ansible task to add another apt repository.  
- With this PR's changes the comments don't get the extraneous space after `#` (note that before changes this behaviour would stop repeating at some point).  

NB: This PR doesn't fix the removal of trailing whitespaces that my org was initially trying to get rid of so idempotency isn't completely fixed as the module keeps modifying .list files unrelated to current task.  

@bcoca does this qualifies as appropriate tests?

bcoca commented:
sounds right, but I'm not sure if we could use some other repo we have more control over, c  @mattclay

pviffx commented:
> we could use some other repo we have more control over  

I don't understand, could you elaborate please?

bcoca commented:
we've had this happen before, for example the repo gets moved or renamed or just removed. I know this is unlikely to happen with docker, but I prefer to err on the safe side.

pviffx commented:
How about Ansible repo as indicated by [installing Ansible on Debian](https://docs.ansible.com/ansible/latest/installation_guide/installation_distros.html#installing-ansible-on-debian)?  

Note that neither Docker or Ansible generate comments in their .list files so in any case adding comments manually is still required for testing.  

Do you want me to write a task in `test/integration/targets/apt_repository/tasks/` for this specific change?

mattclay commented:
I added a test for the comment preservation behavior in https://github.com/ansible/ansible/pull/84904. You can update the assertion in that test as part of this PR to reflect the corrected behavior:

https://github.com/ansible/ansible/blob/87422526f53993e10dd0cfcf465f1d7674357ca5/test/integration/targets/apt_repository/tasks/apt.yml#L57-L61

pviffx committed:
Update apt_repository comment preservation test.

Integration test update following fix in 7f2ea1ffad.

Thanks to Matt Clay for writing the tests.

pviffx commented:
@mattclay thanks **a lot** for the tests!  

After changes introduced in 7f2ea1ffad the test failed until updated in e1af12653a.

webknjaz commented:
/azp run

pviffx commented:
Is there still something needed to have this PR merged? I'd like to not see it die like the original.

Akasurde commented on lib/ansible/modules/apt_repository.py:
```diff
@@ -362,10 +360,10 @@ def save(self):
                 for n, valid, enabled, source, comment in sources:
```
I would combine the for loop https://github.com/ansible/ansible/blob/bb30a7fa1fafc8cd4dc0ff6c509cbbe524319d98/lib/ansible/modules/apt_repository.py#L340 with https://github.com/ansible/ansible/blob/bb30a7fa1fafc8cd4dc0ff6c509cbbe524319d98/lib/ansible/modules/apt_repository.py#L375 into a separate function and do the change on top of it.

Akasurde commented on lib/ansible/modules/apt_repository.py:
```diff
@@ -296,14 +296,12 @@ def _parse(self, line, raise_if_invalid_or_disabled=False):
+        source = line
         if source:
-            chunks = source.split()
+            chunks = re.split(' ', source, maxsplit=1)
```
I am replacing the `_parse` function altogether in https://github.com/ansible/ansible/pull/85881/ so this change is not required.
